extensions/interactions/TextInput/TextInputPredictionServiceSpec.js

```diff
@@ -47,6 +47,7 @@ describe('Text Input Prediction Service', function() {
         for(var j = 0; j < trainingData[i].answers.length; j++) {
           predictedAnswerGroup = predictionService.predict(
             classifierData, trainingData[i].answers[j]);
+          // eslint-disable-next-line eqeqeq  
```

Comment:
Why?

Reply:
On changing `==` to `===` on the next line, one of the tests was failing:

![screenshot from 2018-01-20 15-07-39](https://user-images.githubusercontent.com/24826041/35182255-0a5db130-fdf7-11e7-8891-4c22dd343c27.png)

So I reverted back and disabled the rule for this line. Thanks!

Reply:
That seems unexpected. Could you confirm with @anmolshkl or @prasanna08 that this behaviour is expected? We should understand the reasoning and add it in as a comment.

(Just "the test fails" isn't enough. We should know why this change causes the test to fail.)

Reply:
(Maybe you can also analyze the code to try and understand why it's failing when == becomes ===.)

Reply:
Okay @seanlip! I will look into it and get back. Thanks!

Reply:
I think the problem lies in the different behavior of the operators: `==` only checks the value of the two operands and performs an implicit type conversion if the types do not match. On the other hand, `===` checks both the type as well as the value and does not perform any type conversion. (As stated on [StackOverflow](https://stackoverflow.com/questions/359494/which-equals-operator-vs-should-be-used-in-javascript-comparisons)). 
In our case, `predictedAnswerGroup` seems to be of type object whereas `trainingData[i].answer_group_index` is a literal of the same value. Although I unable to confirm their types yet. @anmolshkl and @prasanna08: Can you please help me with this?  
Thanks!

Reply:
@apb7 the predictedAnswerGroup is being returned from the `SVMPredictionService`. Could you please do a quick check (using typeof) to see the type of the predicted answer group (it is being returned from an array called `classes` from within `SVMPredictionService`).

Reply:
Hello @anmolshkl! It shows two types: number and string, with string being displayed most of the time.

Reply:
@apb7 ummm, you mean to say the type of the same variable varies between string and number? OR that the returned value is a string and the value against which we are checking the returned value is number?

Reply:
The type of the same variable varies between string and number.

Reply:
To clarify, we're looking for cases in which the types of the two things being compared differ.

@apb7 if you could provide repro instructions or a chain of reasoning to explain how that can happen, that would be useful. E.g. can you show the values and types for `predictedAnswerGroup` and `trainingData[i].answer_group_index` that differ, and explain when and why that happens?

Reply:
Okay @seanlip! I am working on this. Thanks!

Reply:
@seanlip: I found out that `trainingData[i].answer_group_index` returns a value of type number as is evident from [text_input_training_data.json](https://github.com/oppia/oppia/blob/develop/core/tests/data/text_input_training_data.json) whereas `predictedAnswerGroup` returns the same value but either of type string or number, which can be confirmed by putting `console.log(prediction);` before this [line](https://github.com/oppia/oppia/blob/develop/extensions/interactions/TextInput/TextInputPredictionService.js#L43) in `TextInputPredictionService.js`.  
Therefore, `==` works but `===` results in a test frontend test failure since the variable `correctPredictions` does not increments and remains zero. Due to which [this] (https://github.com/oppia/oppia/blob/develop/extensions/interactions/TextInput/TextInputPredictionServiceSpec.js#L56) test fails.  
Thanks!

Reply:
OK, thanks @apb7. Can you figure out why predictedAnswerGroup sometimes returns a value of type string? I think you may have found a bug and we need to fix that.

/cc @anmolshkl @prasanna08

Reply:
Okay, I'll figure it out @seanlip! Thanks!

Reply:
@seanlip: On trying `console.log(prediction);`, I found out that predictedAnswerGroup returns a value -1 as type number and all other values as type string. I think this occurs due to some combination of parameters of Support Vector Machine as stated in `text_input_classfier_data.json` although I am not sure about which parameters cause this issue.  
Thanks!

Reply:
OK. I'm not very familiar with this code, but the issue does need to be resolved. You might want to dig deeper and/or work with @anmolshkl and @prasanna08 to address this.

Reply:
Okay, @seanlip! Should we create another issue for this? Also, shall I, for time being, disable `eqeqeq` rule for this line and make other review changes as suggested for the PR?  
Thanks!

Reply:
No, this issue should be resolved in some form before this PR goes in. Otherwise, from past experience, it's basically going to stay there forever.

Reply:
Okay! This sounds much better :) I'll discuss this with @prasanna08 and @anmolshkl. Thanks!

Reply:
Hello @prasanna08 and @anmolshkl! Can you please help me with this issue?  
Thanks!

Reply:
@apb7 it seems that error is because of [classes](https://github.com/oppia/oppia/blob/develop/extensions/classifiers/SVMPredictionService.js#L63) variable in SVMPredictionService.js file and [default prediction](https://github.com/oppia/oppia/blob/develop/extensions/interactions/TextInput/TextInputPredictionService.js#L32) value in text input prediction service. 

So the SVM Prediction data comes from Oppia-ml directly where it is stored from a trained model from scikit learn. I can think following is the cause of this issue:
When you load JSON file, all the values are converted and loaded as a string by default. So the `prediction` variable is bound to have two types as default prediction is a number (-1) but the loaded model will return a string (since it comes from a JSON file).

But the weird thing is this
>  I found out that trainingData[i].answer_group_index returns a value of type number as is evident from text_input_training_data.json 

This should not be the case because trained data is also gets loaded from JSON and it should also return strings rather than numerical values. Can you confirm this?

So the proper fix would be to convert default prediction value to `prediction = '-1'`. This is because when we will share data between Oppia-ml and Oppia over the internet we will be converting train model's data to JSON on Oppia-ml and load it from JSON on Oppia. So we will be playing with strings the whole time.

P.S: I think we have encountered this issue of JSON converting everything to string, before as well. And I remember adding some checks for this conversion on Oppia-ml as per @seanlip's suggestions.

Reply:
@prasanna08 I looked at text_training_input_data.json only. I was not aware of the fact that the JSON values are converted to string by default. Also, on changing the value of prediction to '-1', the frontend tests still fail, citing the same reason.  
Thanks!

Reply:
So it fails with same reason. That's interesting. Can you check the type of prediction returned by TextInputPredictionService again? Previously, as you said, it had two types, a string, and a number. So now it should be a string only. If that is correct then can you check the type of training data's prediction attribute. That should be a string, too, unless `getJSONFixture()` does not work same as python's 'json.loads()' function and if it is the case then covert the trainning data's prediction attribute toString(). I am sure that in Python json.loads() loads everything in Unicode encoding and because of that, we should be dealing with strings all the possible places.

Reply:
Okay, thanks! I'll try this out.

Reply:
@prasanna08: I found out that training data's prediction attribute is a number, that is, `trainingData[i].answer_group_index` is of type `number` whereas `predictedAnswerGroup` is of type String. As suggested, I have used toString() to resolve this problem:  
`if(predictedAnswerGroup === trainingData[i].answer_group_index.toString()){`  
@seanlip: Shall I go ahead with the review changes now?  
Thanks!

Reply:
Done, thanks!

Reply:
I don't like the sound of this. We shouldn't be doing type conversions willy-nilly, that introduces lots of potential for error.

Is the issue that getJSONFixture() converts ints to strings? Have you verified that this is actually the case?

Reply:
Okay, @seanlip! These are my findings:  
* getJSONFixture() does not convert ints to strings. This is the output I get on console.log of `trainingData` in `TextInputPredictionServiceSpec.js`:  
```javascript
[Object{answer_group_index: 1, answers: [..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ...]}, Object{answer_group_index: 2, answers: [..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ..., ...]},
```  
It clearly shows that `answer_group_index` is of type number.  
*  On reverting back to the original state and converting all `answer_group_index` in `base/core/tests/data/text_input_training_data.json` from number to string, the problem is resolved.  
Therefore I suspect that the problem lies in the JSON file itself.  

@prasanna08: 
>This should not be the case because trained data is also gets loaded from JSON and it should also return strings rather than numerical values.
  
This is not happening here.  
Thanks!

Reply:
Ah ok, good investigation, thanks @apb7!

@prasanna08 @anmolshkl can you validate this please? You're both the experts on Oppia's ML stuff so I think at least one of you should verify this is correct before @apb7 proceeds.

Reply:
Seems like getJSONFixture() does not work same as json.loads() in python. It is difference between two platforms. Python considers/converts everything a Unicode when it dumps or loads JSON. That is not the case with getJSONFixture(). It seems that there is no universal rule to the interpretation of JSON file.

Reply:
@seanlip: What do you suggest? We have two options here (as of now):  
* Use the `toString()` method and convert int to string.  Or
* Change the `answer_group_index` from `1` to `"1"` (for example) directly in the JSON file.  
Both of these work.   
@prasanna08: I have a doubt regarding the second approach: Is changing directly in JSON file feasible? Can it affect any other function based on this dataset?  
Thanks!

Reply:
'.toString()' is fine. Since this is just for the test of prediction API. It does not affect actual operation of the system.

Reply:
Thank you, @prasanna08!

Reply:
@prasanna08 `classes` will be a list of string type, right? (since json.dumps on Oppia-ml will dump ints as strings). In that case, the default prediction made by TextInputPredictionService should not be an integer (perhaps null?).

Reply:
@seanlip: As @prasanna08 stated:  
>Since this is just for the test of prediction API. It does not affect actual operation of the system.  

We can also go with the latter approach. What should be done then?  
Thanks!

Reply:
In general, all code in Oppia, whether for tests or production, should adhere to the same consistent standards.

I still do not understand two things:
- predictedAnswerGroup returns a value -1 as type number and all other values as type string. If 
"everything is converted to a string", why is -1 preserved as a number?
- Python considers/converts everything a Unicode when it dumps or loads JSON -- this is not consistent with my experience. The only JSON issue I'm aware of is not being able to pass an OrderedDict. @prasanna08 could you provide a reference to the case you're citing from earlier (or any documentation, or a quick sample test, that would show this is the case)?

Reply:
@seanlip: For the former problem, `predictedAnswerGroup` has  a default value set to `-1` and not `"-1"` in [TextInputPredictionService.js](https://github.com/oppia/oppia/blob/develop/extensions/interactions/TextInput/TextInputPredictionService.js#L32)  whereas the data from which it derives its values, the SVM `classes` in `/core/tests/data/text_input_classifier_data.json` are all strings, ranging from "1" to "5" (which is the value obtained on console.log of predictedAnswerGroup).

Reply:
Ah ok, thanks, that's a good explanation of the discrepancy.

For the second, I ran the following in Python:

```
>>> import json
>>> json.dumps(2)
'2'
>>> json.loads('2')
2
>>> type(json.loads('2'))
<type 'int'>
>>> json.dumps('2')
'"2"'
>>> json.loads('"2"')
u'2'
>>> type(json.loads('"2"'))
<type 'unicode'>
```

so that's why I'm still confused. Did I misinterpret something?

Reply:
Hi @seanlip, my earlier comment was wrong. So JSON converts keys of a dict to strings and not everything. Sorry for that. Actually, I thought that JSON converts everything to strings but then I went back to Oppia-ml's classifiers' validation code and realized that it only converts KEYs to unicode and not other things.

Now I have identified the issue here, which has to do with 'text_input_classifier_data.json' file. See when this file was encoded and stored in JSON all the members of the list `data['SVM']['classes']` were converted to `unicode` before the encoding it into JSON file. So SVMPredictionService returns a member from this list and that is why it returns string.

I also cross verified this with 'code_classifier_data.json' (which stores similar data as 'text_input_classifier_data.json' but for code classifier). Now when you load this file, member of the list `data['SVM']['classses']` are still integers and so SVMPredictionService returns a number. The list should store integers and not strings, that is expected behavior, as these predictions serve as the index to answer_groups of an exploration and an index is a number and not a string.

So the issue is with the encoding of text classifier's data which should have been in integer format before encoding. So we will have to convert it. So after changing 'text_input_classifier_data.json' this PR can be fixed.

Also my test for code classifier (in the CodeReplPredictionServiceSpec.js) is different from @anmolshkl's test in TextClassifierPredictionServiceSpec.js. For instance, I am checking for correctness of the code, that is to verify that the frontend prediction service gives same prediction as the prediction classifier implemented on Oppia-ml. But Anmol's test just runs training data on the algorithm and ensures that the final accuracy is more than 85%. So my test does not verify accuracy explicitly and Anmol's test does not ensure correctness of the frontend API explicitly.

Also one more issue is the different data structures used by  'code_classifier_test_set.json' and 'text_input_training_data.json' (both files are storing the training data inputs along with their classes so that we can cross verify frontend prediction API). For e.g, structure used in 'code_classifier_test_set.json' is (this is prediction results returned by Oppia-ml's code-classifier),
```
{
  'class': ['answer' for answer in answers where answer has 'class' as its prediction class] for class in all classes
}
```
while 'text_input_training_data.json' stores data in following format (this is training data which is used to train classifier):
```
[
  {
    'answer_group_index': answer_group,
    'answers': list of answers in that answer group.
  }
]
```

So I have identified following issues:
* Issue with 'text_input_classifier_data.json' file. This file needs to be resaved after converting data['SVM']['classes'] list members to integer.
* Issue with different kind of tests for frontend prediction API (accuracy vs correctness of algorithm). We have to figure out what to do these?
* Issue with different encoding used in 'code_classifier_test_set.json' and 'text_input_training_data.json'. We can work on this once we have figured out point 2.

So 1st point is the only one that needs to be resolved to fix this issue. 2nd and 3rd points are not related to eqeqeq rule but are part of ML project. So they should not interfere with this PR.

Reply:
That makes much more sense. Thanks for investigating, @prasanna08!

Could you do a PR to fix the first point so that @apb7 can get this in?

For the second point I think we should have both kinds of tests for both algorithms (might as well, right)? And for the third point I agree that we should have a consistent representation.

Reply:
Thank you @prasanna08!  
@seanlip: I tried out the changes as suggested. Just changing the data['SVM']['classes'] to integer in `text_input_classifier_data.json` fixes the problem. Shall I incorporate them into this PR and fix it?

Reply:
Hi @apb7 you will have to modify the source encoding stored in `text_input_classifier_data.json` and not run code after loading and convert it to int. So if you are OK with that then go ahead, otherwise, I will open a new PR and merge it. Thanks.

Also @seanlip I will create new issue for 2nd and 3rd points. Thanks

Reply:
Hey @prasanna08! I did the same: modified the JSON file directly.

Reply:
Hi @apb7:

> I tried out the changes as suggested. Just changing the data['SVM']['classes'] to integer in text_input_classifier_data.json fixes the problem. Shall I incorporate them into this PR and fix it?

That's up to you and @prasanna08. If you do it and @prasanna08 approves your change, then I'm happy :)

Reply:
Hi @apb7 go ahead with changes. I will take a look after that. Thanks